File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    private HashSet<List<Integer>> res = new HashSet<>();
    private int[] nums;
    private int target;

    private void generateSum(List<Integer> path, int sum, int idx){
        if(sum == target){
            res.add(path);
            sum = 0;
            return;
        }

        if(idx >= nums.length){
            return;
        }

        List<Integer> temp = new ArrayList<>(path);
        temp.add(nums[idx]);
        sum += nums[idx];

        if(sum>target){
            sum = 0;
            return;
        }

        generateSum(temp, sum, idx);
        generateSum(temp, sum, idx+1);

        sum -= nums[idx];
        generateSum(path, sum, idx+1);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> path = new ArrayList<>();
        Arrays.sort(candidates);
        this.target = target;
        this.nums = candidates;

        generateSum(path, 0, 0);
        return new ArrayList<>(res);
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        n = len(candidates)
        res = set()
        
        def backtrack(idx, path, sumSoFar):
            if idx == n:
                return 

            if sumSoFar == target:
                res.add(tuple(path))
                return

            if sumSoFar > target:
                return

            #Pick
            backtrack(idx, path + [candidates[idx]], sumSoFar+candidates[idx])
            backtrack(idx + 1, path + [candidates[idx]], sumSoFar+candidates[idx])

            #Not Pick
            backtrack(idx + 1, path, sumSoFar)
        
        backtrack(0, [], 0)
        return list(res)